kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableChangesUtils.java

```diff
@@ -0,0 +1,192 @@
+        if (!protocolVector.isNullAt(rowId)) {
+          Protocol protocol = Protocol.fromColumnVector(protocolVector, rowId);
+          TableFeatures.validateKernelCanReadTheTable(protocol, tablePath);
+        }
+      }
+    }
+
+    // Drop columns if not requested
+    ColumnarBatch result = batch;
+    if (shouldDropProtocolColumn && protocolIdx >= 0) {
+      result = result.withDeletedColumnAt(protocolIdx);
+    }
+
+    int commitInfoIdx = result.getSchema().indexOf("commitInfo");
+    if (shouldDropCommitInfoColumn && commitInfoIdx >= 0) {
+      result = result.withDeletedColumnAt(commitInfoIdx);
+    }
+
+    return result;
+  }
+
+  /**
+   * Adds version and timestamp columns to a columnar batch.
+   *
+   * <p>The version and timestamp columns are added as the first two columns in the batch.
+   *
+   * @param engine the engine for expression evaluation
+   * @param batch the original batch
+   * @param version the version value to add
+   * @param timestamp the timestamp value to add
+   * @return a new batch with version and timestamp columns prepended
+   */
+  public static ColumnarBatch addVersionAndTimestampColumns(
+      Engine engine, ColumnarBatch batch, long version, long timestamp) {
+    StructType schemaForEval = batch.getSchema();
+
+    ExpressionEvaluator commitVersionGenerator =
+        wrapEngineException(
+            () ->
+                engine
+                    .getExpressionHandler()
+                    .getEvaluator(schemaForEval, Literal.ofLong(version), LongType.LONG),
+            "Get the expression evaluator for the commit version");
+
+    ExpressionEvaluator commitTimestampGenerator =
+        wrapEngineException(
+            () ->
+                engine
+                    .getExpressionHandler()
+                    .getEvaluator(schemaForEval, Literal.ofLong(timestamp), LongType.LONG),
+            "Get the expression evaluator for the commit timestamp");
+
+    ColumnVector commitVersionVector =
+        wrapEngineException(
+            () -> commitVersionGenerator.eval(batch), "Evaluating the commit version expression");
+
+    ColumnVector commitTimestampVector =
+        wrapEngineException(
+            () -> commitTimestampGenerator.eval(batch),
+            "Evaluating the commit timestamp expression");
+
+    return batch
+        .withNewColumn(0, new StructField("version", LongType.LONG, false), commitVersionVector)
+        .withNewColumn(
+            1, new StructField("timestamp", LongType.LONG, false), commitTimestampVector);
+  }
+
+  /**
+   * Flattens an iterator of CommitActions into an iterator of ColumnarBatch, adding version and
+   * timestamp columns to each batch.
+   *
+   * @param engine the engine for expression evaluation
+   * @param commits the iterator of CommitActions to flatten
+   * @return an iterator of ColumnarBatch with version and timestamp columns added
+   */
+  public static CloseableIterator<ColumnarBatch> flattenCommitsAndAddMetadata(
```

Comment:
I wonder if this would be simpler if we added a util for `CloseableIterator<CloseableIterator<T>>::flatMap` (or something along those lines, maybe supplier)? With unit tests. Then this code should be easier, and I think we could reuse this elsewhere. I really don't like us re-implementing common iterator concepts like flatMap in multiple places, seems dangerous for little bugs